Benchmark test for callback unary gRPC

mhaidrygoog opened the pull request:
The benchmark can be run using variable CALLBACK_CLIENT as client type in scenarios_json. The code currently only supports UNARY RPCs and will only work with inproc

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,588      Total (=)      1,964,588

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,886,945      Total (>)     10,886,943

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

vjpai commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,125 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ =
+        config.outstanding_rpcs_per_channel() * config.client_channels();
```
Why have so many threads for an async benchmark? The callback RPCs are nonblocking, so can't we just get by with far fewer threads (just like in the existing async QPS test)?

grpc-testing commented:
```
[trickle] No significant performance differences
```

vjpai commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,125 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+  void ThreadFunc(size_t thread_idx, Thread* t) override {
```
Ok, I see that you tried to pigeonhole this into ThreadFunc like the sync API benchmark, but could you instead make the code based on the async API benchmark, just without completion queues etc? Because as it stands, the ThreadFunc is just going to return quickly and then get reinvoked and you're going to have a bunch of active RPCs going on at once. So this won't be ping-pong at all in any case, but will be more like unbounded (unlike the existing models which bound the number of RPCs that are active at once)

mhaidrygoog replied:
I see. Thanks Vijay. I guess I need to clarify my understanding of when a RPC is terminated while using Callback API. Will meet up with you discuss about it. Also will study async client more to address the issues raised in this review.

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,580      Total (=)      1,964,580

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,886,951      Total (=)     10,886,951

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

mhaidrygoog commented:
Based on the concerns raised in the review I have changed the code to make the thread wait for the completion of the previous RPC callback before starting a new one. This is achieved by maintaining a conditional variable per thread. A thread will wait after launching a callback  and is woken up once the callback is done.

Also I'm trying to create a structure similar to ClientRpcContextImpl AsyncClient to make it a go to point for allocations. I haven't figured out how to make it maintain the context and response for the RPC though without it going out of scope so I'll circle around later to work on it.

Also if this looks good next step here would be to attempt to create a state machine similar to the async client,

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,464      Total (<)     10,887,471

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,468      Total (<)     10,887,471

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

vjpai commented:
I'll review the code at some point soon, but can you actually add test scenarios for this so that it becomes part of the test suite? You'll need to edit `tools/run_tests/performance/scenario_config.py` and regen projects.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,473      Total (>)     10,887,470

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

mhaidrygoog commented:
Changed code logic to Issue new a RPC from the callback of the previous RPC. The thread issuing the RPC waits on a conditional variable and is notified by the callback of the last RPC after the thread is marked complete. All RPC threads then notify the main thread which then cleans up the test and exits. Also added a unit test to run the callback client inproc

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,062      Total (<)     10,887,065

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

mhaidrygoog commented:
The unit test is complaining that I didn't check in particular autogenerated files

"qps benchmark scenarios have been updated, please commit test/cpp/qps/qps_json_driver_scenarios.bzl and/or test/cpp/qps/json_run_localhost_scenarios.bzl"

Looking into it

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,060      Total (<)     10,887,066

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

mhaidrygoog commented:
Added auto generated files required to run benchmark unit test for callback

vjpai commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
+    ctxs_[thread_idx]->rpc_done_ = true;
+    thread_cv->notify_one();
```
Likewise why not directly use the original `cv` struct from the `ctx` rather than going through an additional pointer

vjpai commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+  bool ThreadFuncImpl(Thread* t, size_t thread_idx) override {
+    auto* stub = channels_[thread_idx % channels_.size()].get_stub();
+    ClientContext* context = new grpc::ClientContext();
+    SimpleResponse* response = new SimpleResponse();
```
Can these be entries in the bigger `ctx` data structure so that you don't need to do separate `new` and `delete` on them?

vjpai commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+  void NotifyThread(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
```
This is an odd construction; why not

``` cpp
std::lock_guard<std::mutex> thread_lock(ctxs_[thread_idx]->thread_mu_)
```

mhaidrygoog replied:
Sure, I'll fix those up.

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+  CallbackClientRpcData()
+      : thread_mu_(), thread_cv_(), rpc_done_(false), alarm_(nullptr) {}
+
+  ~CallbackClientRpcData() { delete alarm_; }
```
If you use unique_ptr, I believe you won't need to explicitly delete alarm.

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+    int num_threads = config.async_client_threads();
+    if (num_threads <= 0) {  // Use dynamic sizing
+      num_threads = cores_;
+      gpr_log(GPR_INFO, "Sizing async client to %d threads", num_threads);
```
Nit: change "async" to "callback" client so that the log statement isn't outputting exactly the same thing as the log statement for the async client in its version of `NumThreads`

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+    IssueCallbackRpc(thread_idx, t);
+    WaitForCompletionOfAllRpcs(thread_idx);
+
+    NotifyMainThread(thread_idx);
```
Nit: How about renaming this to `NotifyMainThreadOfThreadCompletion` or something that indicates what we're notifying the main thread about?

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
```
Are there some lock annotations you can add to state what this mutex is supposed to be guarding? It's currently not guarding all the variables in this structure because `ctxs_` is not protected by it. Thus it helps to be clear about where it needs to be used.

mhaidrygoog replied:
Sure, its preventing the main thread that spawns the worker threads from ending it. I'll add comments

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+    while (!ctxs_[thread_idx]->rpc_done_) {
+      thread_cv->wait(thread_lock);
+    }
+    ctxs_[thread_idx]->rpc_done_ = false;
```
Why does this need to be set to false here?

mhaidrygoog replied:
yeah I was reusing this flag on my previous design. Will remove.

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+  }
+
+  void WaitForCompletionOfAllRpcs(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
```
See comment in `NotifyThread` about accessing the mutex and condition variable. It applies here as well.

mhaidrygoog replied:
Yeah, I thought I was achieving 'readability' there. Nevermind

hcaseyal replied:
I think once thread_mu is renamed, it will be more readable this way :)

hcaseyal commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+      thread_cv->wait(thread_lock);
+    }
+    ctxs_[thread_idx]->rpc_done_ = false;
+    thread_lock.unlock();
```
Since you're using unique_lock, I don't think there's any need to explicitly unlock it at the end of this function.

mhaidrygoog replied:
Interesting. I'm not extremely familiar with C++ locks but I assumed the same and was hitting TSAN issues without the explicit lock. Will revisit this

hcaseyal replied:
Interesting. If you needed to explicitly unlock it here, I would expect you would have needed to unlock the unique_lock in this function as well: https://github.com/grpc/grpc/blob/475dc6dc5eaf09e1f9af59e3346d937d3b7bbc2e/test/cpp/qps/client_async.cc#L1063

hcaseyal commented on tools/run_tests/performance/scenario_config.py:
```diff
@@ -252,6 +252,19 @@ def scenarios(self):
             categories=[SMOKETEST] + [INPROC] + [SCALABLE])
 
+        yield _ping_pong_scenario(
+            'cpp_protobuf_callback_unary_1channel_100rpcs_1MB',
```
Oh, I see you're sending 1MB messages here. Is that intended? Do we want to have a scenario for 1 channel 1 byte as well?

mhaidrygoog replied:
Not really, just aped the unit test for async. I''ll add a 1 byte 1 channel scenario. I do have a 1 byte 1 channel scenario for the benchmark run. Though that code sits in google3 and I'll send that for review once these changes are approved and integrated

hcaseyal replied:
No need to add it here if async doesn't do it.

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
 }
 
+struct CallbackClientRpcData {
+  CallbackClientRpcData()
```
I was avoiding renaming it to context as we have a seperate ClientContext structure and this particular structure doesn't embed it like the Async test, hence this name to avoid any such assumption. I don't feel strongly about it though and will rename it if this justification does not seem valid.

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+struct CallbackClientRpcData {
```
I guess it should, I'll make the change.

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+
+  ~CallbackClientRpcData() { delete alarm_; }
+
+  std::mutex thread_mu_;
```
sure

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+
+          if (ThreadCompleted()) {
+            // Notify thread of completion
+            NotifyThread(thread_idx);
```
Agreed. I'll make the change

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+    ctxs_[thread_idx] = new CallbackClientRpcData;
+    IssueCallbackRpc(thread_idx, t);
+    WaitForCompletionOfAllRpcs(thread_idx);
+
```
Will do

mhaidrygoog commented on test/cpp/qps/client_async.cc:
```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+  }
+
+  void IssueCallbackRpc(size_t thread_idx, Thread* t) {
+    if (!InitThreadFuncImpl(thread_idx)) {
```
So InitThreadImpl is just a function from client.h that we override. It really does not involve in the synchronization. The only place the rpc_done_ is set true is under the final callback when ThreadCompleted is true. This variable as you mentioned earlier is incorrectly named and should be more on the lines of thread_done_. My upcoming commit however should hopefully remove the need of using this.

hcaseyal replied:
It seems to me like it is involved in synchronization by nature of being called in this method. `IssueCallbackRpc` is called by `ThreadFunc`. Let's say at some point we decide to implement InitThreadFuncImpl to return false sometimes, and it happens to return false when `ThreadFunc` calls it. Then `ThreadFunc` will call `WaitForCompletionOfAllRpcs`, which will wait on the condition: `!ctxs_[thread_idx]->rpc_done_`. But that condition is never true, because "the only place the rpc_done_ is set true is under the final callback when ThreadCompleted is true": ThreadFuncImpl is never called, so the final callback, or indeed any callback, never happens. Does that make sense?

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,892      Total (=)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,714      Total (>)     10,887,712

 No significant differences in binary sizes


```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,892      Total (=)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,718      Total (<)     10,887,722

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,892      Total (=)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,714      Total (>)     10,887,713

 No significant differences in binary sizes


```

mhaidrygoog commented:
Was able to deflake rest of the tests but the grpc_cpp_tsan has been pretty flaky recently -
https://fusion.corp.google.com/projectanalysis/current/KOKORO/prod%3Agrpc%2Fcore%2Fmaster%2Flinux%2Fsanitizer%2Fgrpc_cpp_tsan?search_pattern=grpc_cpp_tsan&project_types=KOKORO&include_inactive_projects=false

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,335      Total (=)      1,965,335

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,225      Total (<)     10,888,230

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,335      Total (=)      1,965,335

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,888,222      Total (<)     10,888,231

 No significant differences in binary sizes


```
